WCSim_fitQun.c

```diff
@@ -0,0 +1,278 @@
+//max_r and max_z is taken to be the outest boundary of the tank
+#define max_r 162.67
+#define max_z 143.542
+#define mu_m 105.66 //mass of muon
+
+R__LOAD_LIBRARY(/opt/WCSim/build/install/lib/libWCSimRoot.so)
+
+//function that perform extraploation
+bool extrapolation(float fq1rpos[2][7][3], float fq1rdir[2][7][3], double entrance[3], double exit[3], double R=max_r, double Z=max_z) {
+    double x = -fq1rpos[0][2][0], y = fq1rpos[0][2][2], z = fq1rpos[0][2][1];
+    double r = sqrt(x*x+y*y);
+    double x_dir = -fq1rdir[0][2][0], y_dir = fq1rdir[0][2][2], z_dir = fq1rdir[0][2][1];
+    double r_dir = sqrt(x_dir*x_dir+y_dir*y_dir);
+    double vt;
+
+    //check whether the vertex position is located in the tank
+    if ((fabs(z)<Z)&&(r<R)) {
+
+        //search for exit point
+        if (z_dir>0) {vt = (Z-z)/z_dir;}
+        else {vt = (-Z-z)/z_dir;}
+        exit[0] = x+x_dir*vt;
+        exit[1] = y+y_dir*vt;
+        exit[2] = z+z_dir*vt;
+        //check if the particle passes through the barrel
+        if ((exit[0]*exit[0]+exit[1]*exit[1])>(R*R)) {
+            double cos_theta = -(x_dir*x+y_dir*y)/(r_dir*r);
+            vt = (r/r_dir)*(cos_theta+sqrt(cos_theta*cos_theta+(R*R)/(r*r)-1));
+            exit[0] = x+x_dir*vt;
+            exit[1] = y+y_dir*vt;
+            exit[2] = z+z_dir*vt;
+        }
+        //search for entrance point
```

Comment:
the code for entrance and exit points are fundamentally the same, try to make a function that can be used in both situations